File: dart_sample/lib/demos/material/badge_tooltip_demo_page.dart
import 'package:material_ui/material_ui.dart';

class BadgeTooltipDemoPage extends StatefulWidget {
  const BadgeTooltipDemoPage({super.key});

  @override
  State<BadgeTooltipDemoPage> createState() => _BadgeTooltipDemoPageState();
}

class _BadgeTooltipDemoPageState extends State<BadgeTooltipDemoPage> {
  int _count = 7;
  bool _isLabelVisible = true;
  bool _useThemeOverrides = false;
  bool _tooltipsVisible = true;
  bool _useRtl = false;

  @override
  Widget build(BuildContext context) {
    Widget content = Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      children: <Widget>[
        const Text('Badge + Tooltip', style: TextStyle(fontSize: 20)),
        const SizedBox(height: 14),
        const Text(
          'Badge geometry plus plain/rich tooltips with hover, directional theming, and custom positioning.',
          style: TextStyle(fontSize: 14, color: Color(0x8A000000)),
        ),
        const SizedBox(height: 14),
        Row(
          children: <Widget>[
            _controlButton('Count +1', () => setState(() => _count++)),
            const SizedBox(width: 8),
            _controlButton(
              _isLabelVisible ? 'Label on' : 'Label off',
              () => setState(() => _isLabelVisible = !_isLabelVisible),
            ),
            const SizedBox(width: 8),
            _controlButton(
              _useThemeOverrides ? 'Theme on' : 'Theme off',
              () => setState(() => _useThemeOverrides = !_useThemeOverrides),
            ),
            const SizedBox(width: 8),
            _controlButton(
              _tooltipsVisible ? 'Tooltips on' : 'Tooltips off',
              () => setState(() => _tooltipsVisible = !_tooltipsVisible),
            ),
          ],
        ),
        const SizedBox(height: 14),
        Row(
          children: <Widget>[
            _controlButton(
              _useRtl ? 'Direction RTL' : 'Direction LTR',
              () => setState(() => _useRtl = !_useRtl),
            ),
          ],
        ),
        const SizedBox(height: 14),
        Container(
          color: const Color(0xFFF7F2FA),
          padding: const EdgeInsets.all(20),
          child: Row(
            mainAxisAlignment: MainAxisAlignment.spaceAround,
            children: <Widget>[
              _probe(
                'Count',
                Badge.count(
                  count: _count,
                  maxCount: 99,
                  isLabelVisible: _isLabelVisible,
                  child: const Icon(Icons.info_outline, size: 32),
                ),
              ),
              _probe(
                'Small',
                Badge(
                  isLabelVisible: _isLabelVisible,
                  child: const Icon(Icons.star_outline, size: 32),
                ),
              ),
              _probe(
                'Scheme tokens',
                Theme(
                  data: Theme.of(context).copyWith(
                    colorScheme: Theme.of(context).colorScheme.copyWith(
                      error: const Color(0xFF00639B),
                      onError: Colors.white,
                    ),
                  ),
                  child: Badge(
                    label: const Text('M3'),
                    isLabelVisible: _isLabelVisible,
                    child: const Icon(Icons.info_outline, size: 32),
                  ),
                ),
              ),
              _probe(
                'Widget override',
                Badge(
                  backgroundColor: const Color(0xFF00695C),
                  textColor: Colors.white,
                  largeSize: 20,
                  offset: const Offset(7, -7),
                  label: const Text('NEW'),
                  isLabelVisible: _isLabelVisible,
                  child: const Icon(Icons.check, size: 32),
                ),
              ),
              _probe(
                _useRtl ? 'Top end RTL' : 'Top end LTR',
                Directionality(
                  textDirection: _useRtl
                      ? TextDirection.rtl
                      : TextDirection.ltr,
                  child: Badge(
                    alignment: AlignmentDirectional.topEnd,
                    label: const Text('END'),
                    isLabelVisible: _isLabelVisible,
                    child: const Icon(Icons.info_outline, size: 32),
                  ),
                ),
              ),
            ],
          ),
        ),
        const SizedBox(height: 14),
        const Text(
          'Hover or long-press these controls:',
          style: TextStyle(fontSize: 14),
        ),
        const SizedBox(height: 14),
        TooltipVisibility(
          visible: _tooltipsVisible,
          child: Directionality(
            textDirection: _useRtl ? TextDirection.rtl : TextDirection.ltr,
            child: Wrap(
              spacing: 12,
              runSpacing: 12,
              children: <Widget>[
                Tooltip(
                  message: 'Default tooltip',
                  child: OutlinedButton(
                    onPressed: () {},
                    child: const Text('Default'),
                  ),
                ),
                Tooltip(
                  richMessage: const TextSpan(
                    children: <InlineSpan>[
                      TextSpan(
                        text: 'Rich ',
                        style: TextStyle(fontWeight: FontWeight.bold),
                      ),
                      TextSpan(text: 'interactive tooltip'),
                    ],
                  ),
                  decoration: const ShapeDecoration(
                    shape: StadiumBorder(),
                    color: Color(0xFF00695C),
                  ),
                  child: OutlinedButton(
                    onPressed: () {},
                    child: const Text('Rich + shape'),
                  ),
                ),
                Tooltip(
                  message: 'Widget override tooltip',
                  preferBelow: false,
                  verticalOffset: 28,
                  decoration: BoxDecoration(
                    color: const Color(0xFF4527A0),
                    borderRadius: BorderRadius.circular(8),
                  ),
                  textStyle: const TextStyle(color: Colors.white, fontSize: 13),
                  waitDuration: const Duration(milliseconds: 250),
                  child: OutlinedButton(
                    onPressed: () {},
                    child: const Text('Above + delay'),
                  ),
                ),
                Tooltip(
                  message: 'Custom right tooltip',
                  positionDelegate: _positionTooltipRight,
                  child: OutlinedButton(
                    onPressed: _noop,
                    child: const Text('Custom right'),
                  ),
                ),
              ],
            ),
          ),
        ),
      ],
    );

    if (!_useThemeOverrides) {
      return content;
    }

    content = BadgeTheme(
      data: const BadgeThemeData(
        backgroundColor: Color(0xFFB3261E),
        textColor: Colors.white,
        largeSize: 18,
        smallSize: 8,
        padding: EdgeInsets.symmetric(horizontal: 5),
        alignment: AlignmentDirectional.bottomEnd,
      ),
      child: TooltipTheme(
        data: TooltipThemeData(
          decoration: BoxDecoration(
            color: const Color(0xFF00695C),
            borderRadius: BorderRadius.circular(6),
          ),
          textStyle: const TextStyle(color: Colors.white, fontSize: 12),
          padding: const EdgeInsetsDirectional.fromSTEB(12, 4, 4, 4),
          waitDuration: const Duration(milliseconds: 150),
          exitDuration: const Duration(milliseconds: 200),
        ),
        child: content,
      ),
    );
    return content;
  }

  Widget _probe(String label, Widget child) {
    return Column(
      mainAxisSize: MainAxisSize.min,
      children: <Widget>[
        child,
        const SizedBox(height: 8),
        Text(label, style: const TextStyle(fontSize: 12)),
      ],
    );
  }

  Widget _controlButton(String label, VoidCallback onPressed) {
    return TextButton(
      onPressed: onPressed,
      style: TextButton.styleFrom(
        backgroundColor: const Color(0xFFEADDFF),
        foregroundColor: const Color(0xFF21005D),
        minimumSize: const Size(0, 36),
      ),
      child: Text(label, style: const TextStyle(fontSize: 12)),
    );
  }

  static Offset _positionTooltipRight(TooltipPositionContext position) {
    return Offset(
      position.target.dx + position.targetSize.width / 2 + 8,
      position.target.dy - position.tooltipSize.height / 2,
    );
  }

  static void _noop() {}
}

File: src/Sample/Plumix.Sample/Demos/Material/BadgeTooltipDemoPage.cs
using System;
using Avalonia;
using Avalonia.Media;
using Plumix.Material;
using Plumix.Rendering;
using Plumix.UI;
using Plumix.Widgets;

// Dart parity source: dart_sample/lib/demos/material/badge_tooltip_demo_page.dart (exact sample parity)

namespace Plumix;

public sealed class BadgeTooltipDemoPage : StatefulWidget
{
    public override State CreateState() => new BadgeTooltipDemoPageState();
}

internal sealed class BadgeTooltipDemoPageState : State
{
    private int _count = 7;
    private bool _isLabelVisible = true;
    private bool _useThemeOverrides;
    private bool _tooltipsVisible = true;
    private bool _useRtl;

    public override Widget Build(BuildContext context)
    {
        var content = new Column(
            crossAxisAlignment: CrossAxisAlignment.Stretch,
            spacing: 14,
            children:
            [
                new Text("Badge + Tooltip", fontSize: 20, color: Colors.Black),
                new Text(
                    "Badge geometry plus overlay tooltips with hover, edge-aware placement, and custom positioning.",
                    fontSize: 14,
                    color: Color.Parse("#8A000000")),
                new Row(
                    spacing: 8,
                    children:
                    [
                        ControlButton("Count +1", () => SetState(() => _count++)),
                        ControlButton(
                            _isLabelVisible ? "Label on" : "Label off",
                            () => SetState(() => _isLabelVisible = !_isLabelVisible)),
                        ControlButton(
                            _useThemeOverrides ? "Theme on" : "Theme off",
                            () => SetState(() => _useThemeOverrides = !_useThemeOverrides)),
                        ControlButton(
                            _tooltipsVisible ? "Tooltips on" : "Tooltips off",
                            () => SetState(() => _tooltipsVisible = !_tooltipsVisible)),
                    ]),
                new Row(
                    children:
                    [
                        ControlButton(
                            _useRtl ? "Direction RTL" : "Direction LTR",
                            () => SetState(() => _useRtl = !_useRtl)),
                    ]),
                new Container(
                    color: Color.Parse("#FFF7F2FA"),
                    padding: new Thickness(20),
                    child: new Row(
                        mainAxisAlignment: MainAxisAlignment.SpaceAround,
                        children:
                        [
                            Probe("Count", Badge.Count(
                                count: _count,
                                maxCount: 99,
                                isLabelVisible: _isLabelVisible,
                                child: new Icon(Icons.InfoOutline, size: 32))),
                            Probe("Small", new Badge(
                                isLabelVisible: _isLabelVisible,
                                child: new Icon(Icons.StarOutline, size: 32))),
                            Probe(
                                "Scheme tokens",
                                new Theme(
                                    Theme.Of(context) with
                                    {
                                        ColorScheme = Theme.Of(context).ColorScheme with
                                        {
                                            Error = Color.Parse("#FF00639B"),
                                            OnError = Colors.White,
                                        },
                                    },
                                    new Badge(
                                        label: new Text("M3"),
                                        isLabelVisible: _isLabelVisible,
                                        child: new Icon(Icons.InfoOutline, size: 32)))),
                            Probe("Widget override", new Badge(
                                backgroundColor: Color.Parse("#FF00695C"),
                                textColor: Colors.White,
                                largeSize: 20,
                                offset: new Vector(7, -7),
                                label: new Text("NEW"),
                                isLabelVisible: _isLabelVisible,
                                child: new Icon(Icons.Check, size: 32))),
                            Probe(
                                _useRtl ? "Top end RTL" : "Top end LTR",
                                new Directionality(
                                    _useRtl ? TextDirection.Rtl : TextDirection.Ltr,
                                    new Badge(
                                        alignment: AlignmentDirectional.TopEnd,
                                        label: new Text("END"),
                                        isLabelVisible: _isLabelVisible,
                                        child: new Icon(Icons.InfoOutline, size: 32)))),
                        ])),
                new Text("Hover or long-press these controls:", fontSize: 14, color: Colors.Black),
                new TooltipVisibility(
                    visible: _tooltipsVisible,
                    child: new Row(
                        spacing: 12,
                        children:
                        [
                            new Tooltip(
                                message: "Default tooltip",
                                child: new OutlinedButton(
                                    onPressed: () => { },
                                    child: new Text("Default"))),
                            new Tooltip(
                                message: "Widget override tooltip",
                                preferBelow: false,
                                verticalOffset: 28,
                                decoration: new BoxDecoration(
                                    Color: Color.Parse("#FF4527A0"),
                                    BorderRadius: BorderRadius.Circular(8)),
                                textStyle: new TextStyle(Color: Colors.White, FontSize: 13),
                                waitDuration: TimeSpan.FromMilliseconds(250),
                                child: new OutlinedButton(
                                    onPressed: () => { },
                                    child: new Text("Above + delay"))),
                            new Tooltip(
                                message: "Custom right tooltip",
                                positionDelegate: position => new Point(
                                    position.Target.X + (position.TargetSize.Width / 2.0) + 8,
                                    position.Target.Y - (position.TooltipSize.Height / 2.0)),
                                child: new OutlinedButton(
                                    onPressed: () => { },
                                    child: new Text("Custom right"))),
                        ])),
            ]);

        if (!_useThemeOverrides)
        {
            return content;
        }

        return new BadgeTheme(
            data: new BadgeThemeData(
                BackgroundColor: Color.Parse("#FFB3261E"),
                TextColor: Colors.White,
                LargeSize: 18,
                SmallSize: 8,
                Padding: new Thickness(5, 0),
                Alignment: AlignmentDirectional.BottomEnd),
            child: new TooltipTheme(
                data: new TooltipThemeData(
                    Decoration: new BoxDecoration(
                        Color: Color.Parse("#FF00695C"),
                        BorderRadius: BorderRadius.Circular(6)),
                    TextStyle: new TextStyle(Color: Colors.White, FontSize: 12),
                    WaitDuration: TimeSpan.FromMilliseconds(150),
                    ExitDuration: TimeSpan.FromMilliseconds(200)),
                child: content));
    }

    private static Widget Probe(string label, Widget child)
    {
        return new Column(
            mainAxisSize: MainAxisSize.Min,
            spacing: 8,
            children: [child, new Text(label, fontSize: 12, color: Colors.Black)]);
    }

    private static Widget ControlButton(string label, Action onPressed)
    {
        return new TextButton(
            onPressed: onPressed,
            backgroundColor: Color.Parse("#FFEADDFF"),
            foregroundColor: Color.Parse("#FF21005D"),
            minHeight: 36,
            child: new Text(label, fontSize: 12));
    }
}
